cargo-marker/src/backend/cargo.rs

```diff
@@ -0,0 +1,78 @@
+use std::process::Command;
+
+use camino::Utf8PathBuf;
+use cargo_metadata::MetadataCommand;
+use serde::Deserialize;
+
+use crate::ExitStatus;
+
+#[derive(Debug, Default)]
+pub struct Cargo {
+    /// The rustc toolchain this driver belongs to. This can be `None` during
+    /// execution commands such as `cargo locate-project`
+    pub(crate) toolchain: Option<String>,
+}
+
+#[derive(Deserialize, Debug)]
+struct ProjectLocation {
+    root: Utf8PathBuf,
+}
+
+impl Cargo {
+    pub fn with_toolchain(toolchain: impl Into<String>) -> Self {
+        Self {
+            toolchain: Some(toolchain.into()),
+        }
+    }
+
+    /// This returns a command calling rustup, which acts as a proxy for the
+    /// Cargo of the selected toolchain.
+    /// It may add additional flags for verbose output.
+    ///
+    /// See also [`super::toolchain::Toolchain::cargo_build_command`] if the
+    /// commands is intended to build a crate.
+    pub fn command(&self) -> Command {
+        // Marker requires rustc's shared libraries to be available. These are
+        // added by rustup, when it acts like a proxy for cargo/rustc/etc.
+        // This also means that cargo needs to be invoked via rustup, to have
+        // the libraries available when Marker is invoked. This is such a mess...
+        // All of this would be so, so much simpler if marker was part of rustup :/
+        if let Some(toolchain) = &self.toolchain {
+            let mut cmd = Command::new("rustup");
+            cmd.args(["run", toolchain, "cargo"]);
+            cmd
+        } else {
+            // for cargo locate-project - this command can be run without
+            // specified toolchain
+            Command::new("cargo")
+        }
+    }
+
+    pub fn cargo_locate_project(&self) -> Result<Utf8PathBuf, ExitStatus> {
+        let mut cmd = self.command();
+
+        let output = cmd
+            .arg("locate-project")
+            .arg("--workspace")
```

Comment:
I worked with @nitkach to make this PR. You are right, the `--wokspace` flag was used to specifically look for a workspace manifest. The current PR was initially intended as more of a refactoring + the usage of `cargo locate-project`. However, this revealed the limitation in the config loading model in the current implementation at least. The current way `cargo-marker` loads the config is that it does it through the workspace metadata only. I think it is a reasonable simplification at least at this stage. The `--workspace` flag also fixes the error you pasted, because it ignores the package manifest.

I think the way how lints are resolved on a package manifest metadata level requires some additional design rigor. This PR won't fix that problem, and doesn't propose a solution for that. It mainly preserves the current assumption that lints are specified at the workspace level. 

For example, I am not at all sure at this point how `cargo-marker` should behave if lints are specified at package manifest level. What if there are lints at both package and workspace manfiest level? Do we want to merge them? Do we want the package manifest override the workspace one? Do we want to have a sepearate lever to toggle that behavior? Do we want to emulate the `{ workspace = true }` inheritance model that cargo uses for dependencies here as well? There are quite a lot of questions to be answered for this design to be settled. I propose to keep things simple yet, and limit users to workspace metadata only at least in the scope if this PR yet.

Reply:
> The `--workspace` flag also fixes the error you pasted, because it ignores the package manifest.

The pasted error came up, after I installed the `cargo-marker` version from this branch. That's why I mentioned it.

> There are quite a lot of questions to be answered for this design to be settled. I propose to keep things simple yet, and limit users to workspace metadata only at least in the scope of this PR yet.

I agree that it's better to limit Marker to workspaces for now :+1:

Reply:
I've looked a bit into the new error:

Marker creates a dummy crate and manifest to fetch all specified lint crates. This manifest should usually only have absolute paths (That's what the `config.normalize()`) call is for in `cargo-marker/src/config.rs`). However, now it seems to use a relative path instead, which seems to be causing the error.